src/rpc/core_rpc_server.cpp

```diff
@@ -2735,52 +2782,60 @@ namespace cryptonote
 
-      const auto proof = m_core.get_uptime_proof(pubkey_info.pubkey);
+      res.service_node_states.push_back(entry);
+    }
 
-      entry.service_node_pubkey           = string_tools::pod_to_hex(pubkey_info.pubkey);
-      entry.registration_height           = pubkey_info.info.registration_height;
-      entry.requested_unlock_height       = pubkey_info.info.requested_unlock_height;
-      entry.last_reward_block_height      = pubkey_info.info.last_reward_block_height;
-      entry.last_reward_transaction_index = pubkey_info.info.last_reward_transaction_index;
-      entry.last_uptime_proof             = proof.timestamp;
-      entry.service_node_version          = {proof.version_major, proof.version_minor, proof.version_patch};
-      entry.public_ip                     = string_tools::get_ip_string_from_int32(pubkey_info.info.public_ip);
-      entry.storage_port                  = pubkey_info.info.storage_port;
+    return true;
+  }
+  //------------------------------------------------------------------------------------------------------------------------------
+  bool core_rpc_server::on_get_n_service_nodes(const COMMAND_RPC_GET_N_SERVICE_NODES::request& req,
+                                               COMMAND_RPC_GET_N_SERVICE_NODES::response& res,
+                                               epee::json_rpc::error&,
+                                               const connection_context*)
+  {
+    std::vector<service_nodes::service_node_pubkey_info> sn_infos = m_core.get_service_node_list_state({});
 
-      entry.contributors.reserve(pubkey_info.info.contributors.size());
+    if (req.limit != 0) {
 
-      using namespace service_nodes;
-      for (service_node_info::contributor_t const &contributor : pubkey_info.info.contributors)
-      {
-        COMMAND_RPC_GET_SERVICE_NODES::response::contributor new_contributor = {};
-        new_contributor.amount   = contributor.amount;
-        new_contributor.reserved = contributor.reserved;
-        new_contributor.address  = cryptonote::get_account_address_as_str(m_core.get_nettype(), false/*is_subaddress*/, contributor.address);
+      const auto limit = std::min(sn_infos.size(), static_cast<size_t>(req.limit));
 
-        new_contributor.locked_contributions.reserve(contributor.locked_contributions.size());
-        for (service_node_info::contribution_t const &src : contributor.locked_contributions)
-        {
-          COMMAND_RPC_GET_SERVICE_NODES::response::contribution dest = {};
-          dest.amount                                                = src.amount;
-          dest.key_image                                             = string_tools::pod_to_hex(src.key_image);
-          dest.key_image_pub_key                                     = string_tools::pod_to_hex(src.key_image_pub_key);
-          new_contributor.locked_contributions.push_back(dest);
-        }
+      std::random_device rd;
+      std::mt19937 mt(rd());
```

Comment:
Calling on a new `std::random_device` on each call is not a good idea (`random_device` can block if you call it enough to exhaust the system's entropy).  Better here to do it just once (per-thread) with:

```C++
static thread_local std::mt19937 mt{std::random_device{}()};
```

Reply:
updated